### PR TITLE
Add foundation for Google Analytics Tracking ID setting

### LINKED ISSUE #3031: Settings Data: Google Analytics
## Feature Description
_Referring to [https://github.com/google/web-stories-wp/issues/2747](https://github.com/google/web-stories-wp/issues/2747)_
Screen Shot 2020-07-06 at 5 25 49 PM

We need an endpoint to read and write google analytics tracking IDs 

## Alternatives Considered
- I did a little discovery and verified with Pascal that this is not something that exists yet that we can access in the Dashboard. 
- SiteKit was mentioned as a possible aid in setting this up 

## Additional Context

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance Criteria

- CRUD functionality for Google Analytics tracking ID:
    - We need to know if the account logged into has a tracking ID set up through an API call 
    - We need to be able to update an existing ID from an API in the dashboard 
    - Ideally as a response we get some error handling returned to us specifying if an update didn't work and why and if it was successful return the new ID back to us with a success message. 
- Availability:
    - ~We need to be able to tell when the account does not have google analytics enabled so that we can disable this form area or provide directions to them to go set it up. ~

## Implementation Brief

- Create a new `Settings` class and initialize it in `Plugin` like the other classes
- Within that class, register a new `web_stories_ga_tracking_id` option that can then be exposed by the `Settings_Controller`
- Create new `Settings_Controller` class inside `includes/REST_API` that exposes a new `web-stories/v1/settings` endpoint.
- This controller would be very similar to `WP_REST_Settings_Controller`, and thus can extend the class, but override `\WP_REST_Settings_Controller::get_registered_options` to only return Web Stories options from the `Settings` class
- Add information about new endpoint to `\Google\Web_Stories\Dashboard::get_dashboard_settings`
- Add new `siteKitSettingsUrl` field to `\Google\Web_Stories\Dashboard::get_dashboard_settings` if Site Kit is installed (defaults to `null` if not installed).

### PR DESCRIPTION
**Pull request by spacedmonkey**
## Summary

Register a custom setting, for web stories analytics. A key was added to the settings endpoint
`wp/v2/settings` under the key name `web_stories_ga_tracking_id`. 

To update this setting, it is simple as. 
```
wp.apiFetch({path: 'wp/v2/settings', method: 'POST', data: { web_stories_ga_tracking_id: 'test' }  });
```

There is also a new capabilities object in the config object. 

`webStoriesDashboardSettings.capabilities.canManageSettings `

UI should disabled / hidden / greyed out if this capability is false. 

## Relevant Technical Choices

Saving this setting, will only be possible for users with `manage_options` capability. 

The REST API currently validate the value passed is a valid string. But there is no server side validation at this time. Any string can be submitted and saved. Validation should happen on the client side. 

Settings object is exposed globally for consistency. 

Settings are only registered on the rest api init, as this follows core function, [register_initial_settings](https://github.com/WordPress/wordpress-develop/blob/e3df1300892ecbbe9a1645d7ce63d1267569bfc8/src/wp-includes/option.php#L1928-L2119) that is hooked in on [rest_api_init](https://github.com/WordPress/wordpress-develop/blob/1285255381f378c4ad4ddbce0a838451760ed499/src/wp-includes/default-filters.php#L455).

I have also preloaded the settings endpoint on the database. This is should not have a lot of overhead to the request. 

## Testing Instructions

---

Fixes #3031

### CODE REVIEW

**Comment by BrittanyIRL** [comment]
I took a bit of time this afternoon to try wiring up the view side of this and it'll work for sure! 

Follow up question is how will 'updating' this work with the added `web_stories_ga_tracking_id` key, what do you want back when we make that request? 

I know it's a proof of concept, but with this how are you planning on implementing permissions to update the id? Since we're supposed to gray out the input with the current ga tracking id when the user's not an admin. 

Thanks for putting this together!

**Comment by spacedmonkey** [comment]
@BrittanyIRL You can see from the screenshot here what the requests look like 
https://github.com/google/web-stories-wp/issues/3031#issuecomment-656115181

As for greying out, that too should be pretty simple. For the stories editor already has a capability object 
https://github.com/google/web-stories-wp/blob/2f8e459ad4edb2cee046b99c8a4db186c29db606/includes/Story_Post_Type.php#L359-L363
passed config object. We could add some capability properties here https://github.com/google/web-stories-wp/blob/2f8e459ad4edb2cee046b99c8a4db186c29db606/includes/Dashboard.php#L236

**Comment by spacedmonkey** [comment]
I have added the cabilibities to the PR See https://github.com/google/web-stories-wp/pull/3098/files#diff-72e6b58089b1f89cc8d9e03231cbb420R243

To access, `capabilities.canManageSettings`

**Comment by BrittanyIRL** [comment]
@spacedmonkey thanks for the additional details, I had missed the bit about `capabilities` in your initial comment.

**Comment by spacedmonkey** [comment]
Do you think this is fit for purpose @BrittanyIRL ? I can continue to work on it to point we can merge. But I think this is mergeable as is.

**Comment by BrittanyIRL** [comment]
> Do you think this is fit for purpose @BrittanyIRL ? I can continue to work on it to point we can merge. But I think this is mergeable as is.

I think that this is great and absolutely a usable direction. 

So long as we have the ability to access the existing google analytics ID, edit the ID, and know if the user has permissions to edit then we're good. (I think you've got all that but just to err on the side of over communication). When you swap this over to ready for review I'll go in and test with endpoint.

File: includes/Settings.php [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+	 */
+	const SETTING_NAME = 'web_stories_ga_tracking_id';
+
+	/**
+	 * Initializes the Settings logic.
+	 *
+	 * @return void
+	 */
+	public function init() {
+		add_action( 'rest_api_init', [ $this, 'register_setting' ] );
```
Comment (swissspidy) [inline includes/Settings.php]: This should run on `init`

File: includes/Settings.php [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+	}
+
+	/**
+	 * Regsiter setting.
+	 *
+	 * @return void
+	 */
+	public function register_setting() {
+		register_setting(
+			'web_stories',
```
Comment (swissspidy) [inline includes/Settings.php]: Maybe `web-stories`? We should strive for some consistency for our names.

File: includes/Settings.php [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+			self::SETTING_NAME,
+			[
+				'show_in_rest' => true,
+				'type'         => 'string',
+				'description'  => __( 'Web stories, GA settings', 'web-stories' ),
```
Comment (swissspidy) [inline includes/Settings.php]:
```suggestion
				'description'  => __( 'Google Analytics Tracking ID', 'web-stories' ),
```

File: includes/Settings.php [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+	/**
+	 * Regsiter setting.
+	 *
+	 * @return void
+	 */
+	public function register_setting() {
+		register_setting(
+			'web_stories',
+			self::SETTING_NAME,
```
Comment (swissspidy) [inline includes/Settings.php]: Add `sanitize_callback`?

File: includes/Settings.php (hunk 1 above)
Comment (spacedmonkey) [reply includes/Settings.php]: In core, [register_initial_settings](https://github.com/WordPress/wordpress-develop/blob/e3df1300892ecbbe9a1645d7ce63d1267569bfc8/src/wp-includes/option.php#L1928-L2119) is hooked in on [rest_api_init](https://github.com/WordPress/wordpress-develop/blob/1285255381f378c4ad4ddbce0a838451760ed499/src/wp-includes/default-filters.php#L455).

Is there any benefit to changing this?

File: includes/Settings.php (hunk 3 above)
Comment (spacedmonkey) [reply includes/Settings.php]: This was just a placeholder.

File: includes/Settings.php (hunk 2 above)
Comment (spacedmonkey) [reply includes/Settings.php]: WordPress normally uses unscores with it's names, see [admin_email](https://github.com/WordPress/wordpress-develop/blob/e3df1300892ecbbe9a1645d7ce63d1267569bfc8/src/wp-includes/option.php#L1981), [default_post_format](https://github.com/WordPress/wordpress-develop/blob/e3df1300892ecbbe9a1645d7ce63d1267569bfc8/src/wp-includes/option.php#L2073) and [use_smilies](https://github.com/WordPress/wordpress-develop/blob/e3df1300892ecbbe9a1645d7ce63d1267569bfc8/src/wp-includes/option.php#L2052). 

Should not follow core coding standards here?

File: includes/Settings.php (hunk 4 above)
Comment (spacedmonkey) [reply includes/Settings.php]: Do we need to make sure it saves in the format of a GA code?

Comment (swissspidy) [reply includes/Settings.php]: We can take a look at validation in a separate issue, and probably even client-side only. Something like `sanitize_text_field` is good enough for now.

File: includes/Plugin.php [hunk 5]
```diff
@@ -73,6 +73,13 @@ class Plugin {
 	 */
 	public $dashboard;
 
+	/**
+	 * Settings.
+	 *
+	 * @var Settings
+	 */
+	public $settings;
```
Comment (swissspidy) [inline includes/Plugin.php:81]: The `Settings` class does not really expose anything useful, so maybe not expose it here for now either?

File: includes/Settings.php (hunk 4 above)
Comment (spacedmonkey) [reply includes/Settings.php]: The REST API already do this validation. 

Screenshot 2020-07-14 at 11 59 12

File: includes/Plugin.php (hunk 5 above)
Comment (spacedmonkey) [reply includes/Plugin.php:81]: It is useful to get the setting name. 

`$plugin->settings:: SETTING_NAME`

File: includes/Settings.php [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+	 */
+	const SETTING_GROUP = 'web_stories';
+
+
+	/**
+	 * Settings handle.
+	 *
+	 * @var string
+	 */
+	const SETTING_NAME = 'web_stories_ga_tracking_id';
```
Comment (swissspidy) [inline includes/Settings.php]: Rename to `SETTING_NAME_TRACKING_ID`. Just `SETTING_NAME` is too generic.

File: includes/Settings.php [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+class Settings {
+
+	/**
+	 * Settings group.
+	 *
+	 * @var string
+	 */
+	const SETTING_GROUP = 'web_stories';
+
+
```
Comment (swissspidy) [inline includes/Settings.php]:
```suggestion

```

Double newlines here

File: includes/Settings.php [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+	public function init() {
+		add_action( 'rest_api_init', [ $this, 'register_setting' ] );
+	}
+
+	/**
+	 * Regsiter setting.
+	 *
+	 * @return void
+	 */
+	public function register_setting() {
```
Comment (swissspidy) [inline includes/Settings.php]:
```suggestion
	public function register_settings() {
```

There's gonna be more in the future

File: includes/Settings.php [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+	 * Initializes the Settings logic.
+	 *
+	 * @return void
+	 */
+	public function init() {
+		add_action( 'rest_api_init', [ $this, 'register_setting' ] );
+	}
+
+	/**
+	 * Regsiter setting.
```
Comment (swissspidy) [inline includes/Settings.php]:
```suggestion
	 * Register settings.
```

File: includes/Plugin.php (hunk 5 above)
Comment (swissspidy) [reply includes/Plugin.php:81]: That's gonna make it hard for us to maintain backwards compatibility though. But well, we can revisit before stable.

File: includes/Dashboard.php [hunk 10]
```diff
@@ -123,6 +123,7 @@ public function load_stories_dashboard() {
 		// TODO Preload templates.
 		$preload_paths = [
 			'/web-stories/v1/fonts',
+			'/wp/v2/settings',
```
Comment (swissspidy) [inline includes/Dashboard.php:126]: Why preload something that might not even be accessed? And is behind a feature flag anyway.

Comment (spacedmonkey) [reply includes/Dashboard.php:126]: What does it hurt to preload these?

Comment (swissspidy) [reply includes/Dashboard.php:126]: Page load time will be (minimally) increased by this, no?

Comment (spacedmonkey) [reply includes/Dashboard.php:126]: I did some testing, it was around 5-15ms. Settings is a pretty little weight REST API request, as all the options are preloaded already. Where as a request to the settings endpoint is around 55ms, as there is a DNS / transport overhead to any http request.

**Comment by spacedmonkey** [comment]
@swissspidy @BrittanyIRL Can please re-review.

**Review by swissspidy — APPROVED** [review]
If this works for Dashboard: 👍

**Comment by BrittanyIRL** [comment]
This works for me! I was able to read and update and set up config in the APP to see the new capabilities object you mentioned in the config. 

One follow up question - which I feel like could be handled separately. 

How do we want to handle error handling here? I was able to set my ID to gibberish `adfadfadfadfadfadf`. 

Shall we vet for the right format for the ID on the front end for starters? 
When should we expect to get errors returned about setting an ID to something that's no good or if there's no analytics set up and nothing to check the input against?

**Review by BrittanyIRL — APPROVED** [review]

**Comment by swissspidy** [comment]
> How do we want to handle error handling here? Shall we vet for the right format for the ID on the front end for starters?

Jonny raised this as well. I think this should be handled in a new ticket to get feedback from UX & product (to clarify the spec).

We could do a simple regex validation on the front end, and _maybe_ the same on the server side, but the latter might not even be needed.

> When should we expect to get errors returned about setting an ID to something that's no good or if there's no analytics set up and nothing to check the input against?

I'm not sure I fully understand the "if there's no analytics set up and nothing to check the input against" part. All we receive from the input is a tracking ID, e.g. `UA-123456789-1`. It would pass regex test, but we don't know whether this is actually a valid ID in Google Analytics. And we can't really know without you having to sign in with Google.

**Pull request merged**